src/java.base/linux/classes/sun/nio/fs/LinuxWatchService.java

```diff
@@ -171,25 +153,94 @@ private static class Poller extends AbstractPoller {
         // sizeof buffer for when polling inotify
         private static final int BUFFER_SIZE = 8192;
 
+        private enum CheckpointRestoreState {
+            NORMAL_OPERATION,
+            CHECKPOINT_TRANSITION,
+            CHECKPOINTED,
+            CHECKPOINT_ERROR,
+            RESTORE_TRANSITION,
+        }
+
         private final UnixFileSystem fs;
         private final LinuxWatchService watcher;
 
         // inotify file descriptor
-        private final int ifd;
+        private int ifd;
         // socketpair used to shutdown polling thread
-        private final int socketpair[];
+        private int socketpair[];
         // maps watch descriptor to Key
         private final Map<Integer,LinuxWatchKey> wdToKey;
         // address of read buffer
         private final long address;
+        private volatile CheckpointRestoreState checkpointState = CheckpointRestoreState.NORMAL_OPERATION;
+
 
-        Poller(UnixFileSystem fs, LinuxWatchService watcher, int ifd, int[] sp) {
+        Poller(UnixFileSystem fs, LinuxWatchService watcher) throws IOException {
+            this.socketpair = new int[2];
+            initFDs();
             this.fs = fs;
             this.watcher = watcher;
-            this.ifd = ifd;
-            this.socketpair = sp;
             this.wdToKey = new HashMap<>();
             this.address = unsafe.allocateMemory(BUFFER_SIZE);
+            jdk.internal.crac.Core.getJDKContext().register(this);
+        }
+
+        private void initFDs() throws IOException {
+            // initialize inotify
+            try {
+                this.ifd = inotifyInit();
+            } catch (UnixException x) {
+                String msg = (x.errno() == EMFILE) ?
+                        "User limit of inotify instances reached or too many open files" :
+                        x.errorString();
+                throw new IOException(msg);
+            }
+
+            // configure inotify to be non-blocking
+            // create socketpair used in the close mechanism
+            try {
+                configureBlocking(ifd, false);
+                socketpair(this.socketpair);
+                configureBlocking(this.socketpair[0], false);
+            } catch (UnixException x) {
+                UnixNativeDispatcher.close(ifd);
+                throw new IOException(x.errorString());
+            }
+        }
+
+        private boolean processCheckpointRestore() throws IOException {
+            if (checkpointState != CheckpointRestoreState.CHECKPOINT_TRANSITION) {
+                return false;
+            }
+
+            synchronized (this) {
+                CheckpointRestoreState thisState;
+                if (wdToKey.size() == 0) {
+                    unsafe.freeMemory(address);
```

Comment:
Although wait, where is the address is reinitialized? That is created in the constructor, and initFDs() reinitializes only file descriptors. This is very likely to be the reason behind those crashes, I can reproduce one locally.

I don't see any reasons to free the memory, it can perfectly be saved to the image. After removing this line, the test has passed 10 of 10 local runs.